core/store/src/trie/ops/iter.rs

```diff
@@ -95,24 +99,66 @@ where
     I: GenericTrieInternalStorage<N, V>,
 {
     /// Create a new iterator.
+    /// If `start` is specified, the iterator will be positioned on the first element
+    /// with key >= `start`, without recording nodes accessed during the seek.
     pub fn new(
         trie_interface: I,
         prune_condition: Option<Box<dyn Fn(&Vec<u8>) -> bool>>,
     ) -> Result<Self, StorageError> {
-        let root = trie_interface.get_root();
-        let mut iter = Self {
+        let iter = Self {
             trail: Vec::with_capacity(8),
             key_nibbles: Vec::with_capacity(64),
+            initialized: false,
             trie_interface,
             prune_condition,
         };
-        iter.descend_into_node(root)?;
         Ok(iter)
     }
 
     /// Position the iterator on the first element with key >= `key`.
     pub fn seek_prefix<K: AsRef<[u8]>>(&mut self, key: K) -> Result<(), StorageError> {
-        self.seek_nibble_slice(NibbleSlice::new(key.as_ref()), true)?;
+        self.seek_nibble_slice(NibbleSlice::new(key.as_ref()), true, AccessOptions::DEFAULT)?;
+        Ok(())
+    }
+
+    /// Position the iterator on the first element with key >= `key`,
+    /// or the first element with key > `key` if `upper_bound` is true.
+    /// Does not record nodes accessed during the seek.
+    pub fn seek<K: AsRef<[u8]>>(&mut self, key: K, upper_bound: bool) -> Result<(), StorageError> {
+        self.seek_nibble_slice(
+            NibbleSlice::new(key.as_ref()),
+            false,
+            AccessOptions::NO_SIDE_EFFECTS,
+        )?;
+
+        // By this point, we are already positioned the iterator such that
+        // next() will return the first element with key >= `key`.
+        // If `upper_bound` is true, we need to check if the next element
+        // is actually key == `key` and skip it.
+        if upper_bound {
+            if let Some(last) = self.trail.last() {
+                let mut compare_with = self.key_nibbles.clone();
+
+                let is_entering_value_node = match (last.status, &last.node) {
+                    (CrumbStatus::Entering, GenericTrieNode::Branch { value: Some(_), .. }) => true,
+                    (CrumbStatus::Entering, GenericTrieNode::Leaf { .. }) => true,
+                    _ => false,
+                };
+
+                // If the iterator is positioned on a leaf, we need to consider
+                // the extension of the leaf as part of the key to compare.
+                if let GenericTrieNode::Leaf { extension, .. } = &last.node {
+                    let existing_key = NibbleSlice::from_encoded(&extension).0;
+                    compare_with.extend(existing_key.iter());
+                }
+
+                if is_entering_value_node
+                    && key.as_ref() == NibbleSlice::nibbles_to_bytes(&compare_with)
+                {
+                    self.iter_step();
```

Comment:
I suspect this should always return `IterStep::Value(_)` and it might be worth asserting on that.